website/mailing_list/utils.py

```diff
@@ -0,0 +1,422 @@
+            'address': email,
+            'subscribed': enabled and email == user.username and user not in get_unsubscribes(node),
+            'vars': json.dumps({'_id': user._id, 'primary': email == user.username or bool(old_email)}),
+            'upsert': True
+        }
+    )
+    if resp.status_code != 200:
+        raise HTTPError(resp.status_code, data={'message_long': resp.json() or ''})
+
+@require_mailgun
+def remove_user_from_list(node_id, user_id):
+    """ Removes single member of a mailing list on Mailgun
+    Called when a contributor is removed from a Node.
+
+    :param str node_id: ID of node in question
+    :param str user_id: ID of user to remove
+    """
+    from website.models import User  # avoid circular import
+    user = User.load(user_id)
+
+    for email in user.emails:
+        resp = requests.delete(
+            '{}/{}/members/{}'.format(MAILGUN_BASE_LISTS_URL, address(node_id), email),
+            auth=('api', settings.MAILGUN_API_KEY)
+        )
+        if resp.status_code not in [200, 404]:
+            raise HTTPError(resp.status_code, data={'message_long': resp.json() or ''})
+
+@require_mailgun
+def update_multiple_users_in_list(node_id, members):
+    """ Adds/updates members of a mailing list on Mailgun
+
+    :param str node_id: The id of the node in question
+    :param list members: List of json-formatted user dicts to add/enable/update
+    """
+    resp = requests.post(
+        '{}/{}/members.json'.format(MAILGUN_BASE_LISTS_URL, address(node_id)),
+        auth=('api', settings.MAILGUN_API_KEY),
+        data={
+            'upsert': True,
+            'members': json.dumps(members)
+        }
+    )
+    if resp.status_code != 200:
+        raise HTTPError(resp.status_code, data={'message_long': resp.json() or ''})
+
+@require_mailgun
+def full_update(node_id):
+    """ Matches remote list with internal representation
+    :param str node_id: The node to update the mailing list for
+    """
+    from website.models import Node  # avoid circular import
+    node = Node.load(node_id)
+
+    info, members = get_list(node_id)
+
+    if node.is_deleted:
+        delete_list(node_id)
+    elif node.mailing_enabled:
+        if 'list' in info.keys():
+            info = info['list']
+            remote_subscribes = [member['address'] for member in members['items'] if member['subscribed']].sort()
+            local_subscribes = [user.username for user in get_recipients(node)].sort()
+
+            if info['name'] != list_title(node):
+                # Update title if necessary
+                update_title(node_id, node.title)
+
+            if remote_subscribes != local_subscribes:
+                # Push local changes
+                members_list = jsonify_users_list(node.contributors, unsubs=get_unsubscribes(node))
+                update_multiple_users_in_list(node_id, members_list)
+
+            # Delete any noncontribs
+            from website.models import User  # avoid circular import
```

Comment:
I'm all for avoiding circular imports, but can we perhaps group all the dependencies at the top of the function?  (node, user, etc)